### DISCUSSION TITLE
building with cmake and iconv

### DISCUSSION THREAD

**Discussion by userdocs** [arvidn/libtorrent#6215] [Q&A] [answered]
I have encountered a few issues with this setup and was wondering if some things could be clarified.

https://github.com/arvidn/libtorrent/blob/RC_1_2/CMakeLists.txt#L677

When building on Alpine i need to set -`D Iconv_LIBRARY="${lib_dir}/libiconv.a"` otherwise the cmake files libtorrent generates tries to load an `so` file which give me a linking error. From the `build/CMakeCache.txt`

`-D Iconv_LIBRARY="/usr/lib/libc.so"`

Regarding ICU

`iconv` has become problematic for a consistent build usage when factoring in `qtbase` and `boost` seem to prefer ICU if it is available. Whilst I am unsure if that means it is actually used by libtorrent, this defaulting to `iconv` behaviour in cmake has created the issue I do not have when using `b2`

Now I know this is not a specific problem of this project, but what I think would make sense for a direct route to build on top of this with qbittorrent would be to support ICU?

<https://bugreports.qt.io/browse/QTBUG-63529> and [specifically this comment](https://bugreports.qt.io/browse/QTBUG-63529?focusedCommentId=373679&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-373679)

<https://bugreports.qt.io/browse/QTBUG-84708> + [the required fix](https://bugreports.qt.io/browse/QTBUG-84708?focusedCommentId=514864&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-514864)

From the linked fix conversation above will allow a locally built `libiconv` to work with qt using `-iconv`

```
QMAKE_CXXFLAGS=-fpermissive
```

https://bugreports.qt.io/browse/QTBUG-84687 

Would it be possible to get an ICU support option to not have this `iconv` dependency with cmake? 

It makes more sense to me to do this as the main deps all have a preference for ICU over iconv.

ICU > boost > libtorrent > qt > qbittorrent

**Comment by userdocs** [comment, marked as answer]
The cmake pkconfig

v1.2.13
```
prefix=/root/qb-build
libdir=${prefix}/lib

Name: libtorrent
Description: Bittorrent library
Version: 1.2.13
Libs: -L${libdir} -ltorrent-rasterbar -lIconv::Iconv -lssl -lcrypto -ldl -ldl -lcrypto
Cflags: -fexceptions -I/root/qb-build/include -I/root/qb-build/boost -I/root/qb-build/include -DBOOST_ASIO_ENABLE_CANCELIO -DTORRENT_USE_ICONV -DTORRENT_USE_OPENSSL -DTORRENT_USE_LIBCRYPTO
```

v2.0.3.0

```
prefix=/root/qb-build
libdir=${prefix}/lib

Name: libtorrent
Description: Bittorrent library
Version: 2.0.3
Libs: -L${libdir} -ltorrent-rasterbar -lssl -lcrypto -ldl -ldl
Cflags: -fexceptions -I/root/qb-build/include -I/root/qb-build/boost -I/root/qb-build/include -DBOOST_ASIO_ENABLE_CANCELIO -DBOOST_ASIO_NO_DEPRECATED -DTORRENT_USE_OPENSSL -DTORRENT_USE_LIBCRYPTO -DTORRENT_SSL_PEERS -DOPENSSL_NO_SSL2
```

Two things

Why do we have a double `-ldl -ldl` - I am not sure we even need `-ldl` here?

v1 has this `-lIconv::Iconv` but v2 does not.

**Reply by arvidn** [reply]
libtorrent 2.x no longer depends on iconv

**Reply by userdocs** [reply]
What does it do instead or is it just not needed to display unicode?

**Reply by arvidn** [reply]
`iconv` was used to convert from UTF-8 to the local locale of the filesystem. On windows `MultiByteToWideChar()` is used directly, and otherwise `wcrtomb()` is used to convert a unicode codepoint into the native locale. Which is extremely likely to be UTF-8 also.
